libvast/src/type.cpp

```diff
@@ -261,18 +309,32 @@ type::type(std::string_view name, const type& nested,
     const auto attributes_offset = [&]() noexcept
       -> flatbuffers::Offset<
         flatbuffers::Vector<flatbuffers::Offset<fbs::type::detail::Attribute>>> {
-      if (attributes.empty())
+      if (attributes.empty() && merged_attributes.empty())
         return 0;
       auto attributes_offsets
         = std::vector<flatbuffers::Offset<fbs::type::detail::Attribute>>{};
-      attributes_offsets.reserve(attributes.size());
+      attributes_offsets.reserve(attributes.size() + merged_attributes.size());
       for (const auto& attribute : attributes) {
         const auto key_offset = builder.CreateString(attribute.key);
         const auto value_offset
           = attribute.value ? builder.CreateString(*attribute.value) : 0;
         attributes_offsets.emplace_back(fbs::type::detail::CreateAttribute(
           builder, key_offset, value_offset));
       }
+      for (const auto& attribute : merged_attributes) {
+        // Skip over any attributes that were already in the new list of
+        // attributes.
+        if (std::any_of(attributes.begin(), attributes.end(),
+                        [&](const auto& new_attribute) noexcept {
+                          return new_attribute.key == attribute.key;
+                        }))
+          continue;
```

Comment:
Have you considered changing the attributes parameter to those constructors to `std::vector<attribute_view>`? Since we're constructing a new flatbuffer here that shouldn't give us lifetime problems.

With that change you can start out with the provided `attributes` in the upper loop and apply the deduplication there, then you only need to deal with the merged attributes for the remainder of the function.

Reply:
We'd probably need both overloads then, as we have `attribute` at many call sizes, and while we can make `attribute` implicitly convertible to `attribute_view` we cannot make that for containers of them. I do agree though that this would be a good API improvement.

Reply:
@tobim kindly provided me with a diff that implements most of this change. I'll incorporate that in a bit.